src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -379,288 +434,369 @@ private static HttpRequestException GetVersionException(HttpRequestMessage reque
+            TimeSpan pooledConnectionLifetime = _poolManager.Settings._pooledConnectionLifetime;
+            if (pooledConnectionLifetime != Timeout.InfiniteTimeSpan)
+            {
+                return connection.GetLifetimeTicks(Environment.TickCount64) > pooledConnectionLifetime.TotalMilliseconds;
+            }
 
+            return false;
+        }
+
+        private static Exception CreateConnectTimeoutException(OperationCanceledException oce)
+        {
+            // The pattern for request timeouts (on HttpClient) is to throw an OCE with an inner exception of TimeoutException.
+            // Do the same for ConnectTimeout-based timeouts.
+            TimeoutException te = new TimeoutException(SR.net_http_connect_timedout, oce.InnerException);
+            Exception newException = CancellationHelper.CreateOperationCanceledException(te, oce.CancellationToken);
+            ExceptionDispatchInfo.SetCurrentStackTrace(newException);
+            return newException;
+        }
+
+        private async Task AddHttp11ConnectionAsync(HttpRequestMessage request)
+        {
+            if (NetEventSource.Log.IsEnabled()) Trace("Creating new HTTP/1.1 connection for pool.");
+
+            HttpConnection connection;
+            using (CancellationTokenSource cts = GetConnectTimeoutCancellationTokenSource())
+            {
+                try
+                {
+                    connection = await CreateHttp11ConnectionAsync(request, false, cts.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException oce) when (oce.CancellationToken == cts.Token)
+                {
+                    HandleHttp11ConnectionFailure(CreateConnectTimeoutException(oce));
+                    return;
+                }
+                catch (Exception e)
+                {
+                    HandleHttp11ConnectionFailure(e);
+                    return;
+                }
+            }
+
+            // Add the established connection to the pool.
+            ReturnHttp11Connection(connection, isNewConnection: true);
+        }
+
+        private void CheckForHttp11ConnectionInjection()
+        {
+            Debug.Assert(HasSyncObjLock);
+
+            if (!_http11RequestQueue.TryPeekNextRequest(out HttpRequestMessage? request))
+            {
+                return;
+            }
+
+            // Determine if we can and should add a new connection to the pool.
+            if (_availableHttp11Connections.Count == 0 &&                           // No available connections
+                _http11RequestQueue.Count > _pendingHttp11ConnectionCount &&        // More requests queued than pending connections
+                _associatedHttp11ConnectionCount < _maxHttp11Connections)           // Under the connection limit
+            {
+                _associatedHttp11ConnectionCount++;
+                _pendingHttp11ConnectionCount++;
+
+                Task.Run(() => AddHttp11ConnectionAsync(request));
```

Comment:
Why does this needs to run completely separately? Wouldn't just `var _ = AddHttp11ConnectionAsync(request)` be enough? Is this about the lock we have here? It would be nice to at least some comment explaining why it needs to run this way, so that someone doesn't remove it accidentally.

Reply:
It's partly about the lock, but that could be handled differently -- we could just make the call outside the lock.

The main issue is that we don't want to do a lot of work on the calling thread here. If we invoke AddHttp11ConnectionAsync here directly (or in the callers), we are potentially doing a lot of work in this call, including calling the user's ConnectCallback etc. We don't want the calling thread here to be tied up doing that. So instead we queue a work item to the thread pool to do whatever work is necessary to create the new connection, while the current thread can continue on doing whatever it was about to do -- whether that is to wait for (any) available connection or to finish up tearing down an existing connection, etc.